enhancements/20240528-migrate-harvester-api-to-standardized-k8s-api.md

```diff
@@ -0,0 +1,159 @@
+harvesterCopyConfig.APIPath = "/apis"
+harvesterCopyConfig.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+harvesterSubresourceClient, err := rest.RESTClientFor(harvesterCopyConfig)
+
+h.harvesterSubresourceClient.Post().Resource("nodes").SubResource("uncordon").Name(node.Name).Do(context.Background())
+h.harvesterSubresourceClient.Post().Namespace("default").Resource("virtualmachines").SubResource("stop").Name("test").Do(context.Background())
+```
+
+Or a pure native k8s API.
+
+```
+https://{server}/apis/subresources.harvesterhci.io/v1beta1/namespaces/default/virtualmachines/test2/stop
+```
+
+### Use Case 2 - Security
+
+When using subresource API, we can leverage the Kubernetes RBAC to control the access to the subresource API. For example, if we create a role like below, the user can only GET and POST the subresource.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: subresource-test
+  namespace: default
+rules:
+- apiGroups:
+  - subresources.harvesterhci.io
+  resources:
+  - virtualmachines/stop
+  verbs:
+  - get
+  - create
+```
+
+It will get the 403 forbidden error if the user tries to access the subresource without the permission.
+
+```json
+{
+  "kind": "Status",
+  "apiVersion": "v1",
+  "metadata": {},
+  "status": "Failure",
+  "message": "virtualmachines.subresources.harvesterhci.io \"test2\" is forbidden: User \"system:serviceaccount:default:test-subresource\" cannot create resource \"virtualmachines/start\" in API group \"subresources.harvesterhci.io\" in the namespace \"default\"",
+  "reason": "Forbidden",
+  "details": {
+    "name": "test2",
+    "group": "subresources.harvesterhci.io",
+    "kind": "virtualmachines"
+  },
+  "code": 403
+}
+```
+
+### Design
+
+In order to implement this proposal, we need to take the following steps:
+
+1. Define the `APIService` to proxy requests to Harvester API server.
+    ```yaml
+    apiVersion: apiregistration.k8s.io/v1
+    kind: APIService
+    metadata:
+      name: v1beta1.subresources.harvesterhci.io
+    spec:
+      group: subresources.harvesterhci.io
+      groupPriorityMinimum: 1000
+      insecureSkipTLSVerify: true
+      service:
+        name: harvester
+        namespace: harvester-system
+        port: 8443
+      version: v1beta1
+      versionPriority: 15
+    ```
+2. Define the new routes path to accept following request path formats:
+   - `/apis/subresources.harvesterhci.io/v1beta1/namespaces/{namespace}/{resource}/{name}/{subresource}` 
+   - `/apis/subresources.harvesterhci.io/v1beta1/{resource}/{name}/{subresource}`
+
+It's basically done because we've had the action business logic in the controller, we just need to create a new interface to fulfill two different API sources, which one is from steve and the other is from subresource api.
+
+### Test Plan
+
+Should test old and new APIs path formats, both should work as expected.
```

Comment:
The test plan should specify more accurately the new API routes and their functions.
Additionally it should specify which old API routes are expected to be removed in the future.
For the time period where both API schemes are available in parallel, the tests would need to ensure that both API schemes work as intended.